File: src/Scripting/Perl/proxy.pl
use strict;
use warnings;

package RemoteDocumentStore;
sub new { bless { fetch_count => 0 }, shift }
sub get {
    my ($self, $id) = @_;
    $self->{fetch_count}++;
    return "doc($id)";
}

package DocumentStoreProxy;
sub new { bless { backend => undef, cache => {} }, shift }
sub get {
    my ($self, $id) = @_;
    return $self->{cache}{$id} if exists $self->{cache}{$id};
    $self->{backend} //= RemoteDocumentStore->new();
    my $value = $self->{backend}->get($id);
    $self->{cache}{$id} = $value;
    return $value;
}
sub backend_count { defined $_[0]{backend} ? 1 : 0 }
sub fetch_count { defined $_[0]{backend} ? $_[0]{backend}{fetch_count} : 0 }

package main;
my $store = DocumentStoreProxy->new();
my $first = $store->get(42);
my $second = $store->get(42);
print 'backend=', $store->backend_count(), ';fetches=', $store->fetch_count(), ";first=$first;second=$second\n";

File: src/Scripting/Ruby/proxy.rb
class RemoteDocumentStore
  attr_reader :fetch_count

  def initialize
    @fetch_count = 0
  end

  def get(id)
    @fetch_count += 1
    "doc(#{id})"
  end
end

class DocumentStoreProxy
  attr_reader :backend

  def initialize
    @backend = nil
    @cache = {}
  end

  def get(id)
    return @cache[id] if @cache.key?(id)

    @backend ||= RemoteDocumentStore.new
    @cache[id] = @backend.get(id)
  end

  def backend_count
    @backend.nil? ? 0 : 1
  end

  def fetch_count
    @backend&.fetch_count || 0
  end
end

store = DocumentStoreProxy.new
first = store.get(42)
second_value = store.get(42)
puts "backend=#{store.backend_count};fetches=#{store.fetch_count};first=#{first};second=#{second_value}"
